### PR TITLE
Allow up/destroy using rundb data

### LINKED ISSUE #447: Allow teardown (destroy) based upon data in rundb
Per discussions with @jaypoulz  and @scoheb, teardowns using dynamic data require the user to remember the data being passed. To that end, a switch to read the stored data from the RunDB on teardown could be useful.

### PR DESCRIPTION
**Pull request by herlo**
Fixes #447.

A user can now repeat any transaction using the 'tx_id' value. An 'up' action can be repeated by either using the 'tx_id' (with or without targets for filtering), or with a run_id and target. A 'destroy' action can be performed similarly. This can be enabled by setting 'use_rundb_for_actions = True' in linchpin.conf

### CODE REVIEW

**Comment by jaypoulz** [comment]
This worked at first, but now I"m seeing:
```
Traceback (most recent call last):
  File "/usr/bin/linchpin", line 11, in <module>
    load_entry_point('linchpin==1.5.2rc1', 'console_scripts', 'linchpin')()
  File "/usr/lib/python2.7/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/usr/lib/python2.7/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/lib/python2.7/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/lib/python2.7/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/click/decorators.py", line 64, in new_func
    return ctx.invoke(f, obj, *args[1:], **kwargs)
  File "/usr/lib/python2.7/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/linchpin/shell/__init__.py", line 443, in journal
    output += 'Action: {0}\n'.format(v['action'])
TypeError: 'NoneType' object has no attribute '__getitem__'
```

This is my generated runDB:
```
{"dummy_target": {"1": {"inputs": [{"topology_data": {"topology_name": "dummy", "resource_groups": [{"resource_group_name": "dummy", "resource_definitions": [{"count": 1, "role": "dummy_node", "name": "web"}], "resource_group_type": "dummy"}]}}, {"layout_data": {"inventory_layout": {"hosts": {"example-node": {"count": 1, "host_groups": ["example"]}}, "vars": {"hostname": "__IP__"}}}}], "end": "03/23/2018 01:33:40 PM", "outputs": [{"resources": [{"failed": false, "changed": true, "hosts": ["web-7eb4-0.example.net"], "dummy_file": "/tmp/dummy.hosts"}]}], "uhash": "7eb4", "start": "03/23/2018 01:33:34 PM", "rc": 0, "action": "up"}}, "linchpin": {"1": {"action": "up", "targets": [{"dummy_target": {"1": {"uhash": "7eb4", "rc": 0}}}]}}}
```

File: linchpin/__init__.py [hunk 1]
```diff
@@ -383,17 +394,86 @@ def _validate_topology(self, topology):
...
 
-    def do_action(self, provision_data, action='up', run_id=None):
+    def get_pf_data_from_rundb(self, targets, run_id=None, tx_id=None):
+        """
+        This function takes the action and provision_data, returns the
+        pinfile data
+
+        :param targets: A list of targets for which to get the data
+
+        :param action: Action taken (up, destroy, etc). (Default: up)
```
Comment (p3ck) [inline linchpin/__init__.py]: I don't see param action anymore...  and it looks to be hard coded to 'up' now.

Comment (herlo) [reply linchpin/__init__.py]: Are you referring to the do_action method? It got moved further down in the code is all.

https://github.com/CentOS-PaaS-SIG/linchpin/pull/512/files/8719bd6245db8a6fa5c6e0df908f65a4a3605e02#diff-eda7cdccb38e3e91756c08ffdbc7ac4e

Line 463 from what I see...

Comment (p3ck) [reply linchpin/__init__.py]: Please update the inline docs for the function then.  Thanks.

**Comment by herlo** [comment]
[test]

**Comment by p3ck** [comment]
[test] internal jenkins bad

**Comment by jaypoulz** [comment]
Pulled down the new code today, and here are some updated results:

Here's a linchpin journal output of my results:
```
Target: dummy_target
run_id	    action	     uhash	        rc	
--------------------------------------------------
4      	  destroy	     b4ad	        0	
3      	  destroy	     b4ad	        0	
2      	       up	     b4ad	        0	
1      	       up	     5697	        0	
```

Both transactions 3 and 4 were called with:
```
linchpin destory -t 1
```

Shouldn't the uhash for them be 5697 instead of the latest target?

Secondly, the error I was seeing has to do with the the view print loop I believe. With my 4 transcations, if I run `linchpin journal --view tx` I get:
```
ID: 4			Action: destroy

Target              	Run ID	uHash	Exit Code 
-------------------------------------------------
dummy_target        	     4	 b4ad	        0

=================================================

ID: 3			Action: destroy

Target              	Run ID	uHash	Exit Code 
-------------------------------------------------
dummy_target        	     3	 b4ad	        0

=================================================

ID: 2			Action: up

Target              	Run ID	uHash	Exit Code 
-------------------------------------------------
dummy_target        	     2	 b4ad	        0

=================================================
```

But if I up the count parameter to 5 (e.g. `linchpin journal --view tx -c 5`, I get:
```
Traceback (most recent call last):
  File "/usr/bin/linchpin", line 11, in <module>
    load_entry_point('linchpin==1.5.2rc1', 'console_scripts', 'linchpin')()
  File "/usr/lib/python2.7/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/usr/lib/python2.7/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/lib/python2.7/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/lib/python2.7/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/click/decorators.py", line 64, in new_func
    return ctx.invoke(f, obj, *args[1:], **kwargs)
  File "/usr/lib/python2.7/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/linchpin/shell/__init__.py", line 443, in journal
    output += 'Action: {0}\n'.format(v['action'])
TypeError: 'NoneType' object has no attribute '__getitem__'
```

**Comment by herlo** [comment]
[merge]

**Pull request merged**